File: webstore/webstore-java/src/main/java/com/renan/webstore/dto/mapper/UserDTOMapper.java
package com.renan.webstore.dto.mapper;

import com.renan.webstore.business.JwtService;
import com.renan.webstore.dto.UserDTO;
import com.renan.webstore.model.AppUser;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.List;
import java.util.function.Function;

@Component
@RequiredArgsConstructor
public class UserDTOMapper implements Function<AppUser, UserDTO> {
    
    private final JwtService jwtService;
    
    @Override
    public UserDTO apply(AppUser user) {
        List<String> roles = getRoles(user.getRoles());
        return new UserDTO(
                user.getEmail(),
                jwtService.generateJWT(user.getId(), user.isActive(), user.getEmail(), user.getFirstName(), user.getLastName(), roles),
                user.getId(),
                user.isActive(),
                jwtService.TIMEOUT,
                user.getFirstName(),
                user.getLastName(),
                roles
        );
    }
    
    List<String> getRoles(String userRoles){
        return Arrays.stream(userRoles.split(",")).toList();
    }
    
}

File: webstore/webstore-kotlin/src/main/kotlin/com/renan/webstore/dto/mapper/UserDTOMapper.kt
package com.renan.webstore.dto.mapper

import com.renan.webstore.business.JwtService
import com.renan.webstore.business.TIMEOUT
import com.renan.webstore.dto.UserDTO
import com.renan.webstore.model.AppUser
import com.renan.webstore.model.STATUS_ACTIVE
import com.renan.webstore.model.buildRoles
import org.springframework.stereotype.Component
import java.util.function.Function

@Component
class UserDTOMapper(
    val jwtService: JwtService
) : Function<AppUser, UserDTO> {

    override fun apply(user: AppUser): UserDTO {

        return UserDTO(
            email = user.email,
            jwt = jwtService.generateJWT(
                idUser = user.id!!,
                active = user.userStatus == STATUS_ACTIVE,
                email = user.email,
                firstName = user.firstName,
                lastName = user.lastName,
                roles = buildRoles(user.roles)
            ),
            id = user.id,
            active = user.userStatus == STATUS_ACTIVE,
            expiresIn = TIMEOUT,
            name = user.firstName,
            lastName = user.lastName,
            roles = buildRoles(user.roles)
        )
    }

}